app/controllers/users/sessions_controller.rb

```diff
@@ -44,19 +55,42 @@ def alive?
       current_user.present? && session_alive
     end
 
-    def track_authentication_attempt(email)
-      user = User.find_by(email: email.downcase) || AnonymousUser.new
+    def existing_user
+      @_existing_user ||= User.find_by(email: params[:user][:email].downcase) || AnonymousUser.new
+    end
 
+    def track_authentication_attempt
       properties = {
-        success?: current_user.present?, user_id: user.uuid
+        success?: current_user.present?, user_id: existing_user.uuid
       }
 
       analytics.track_event(Analytics::EMAIL_AND_PASSWORD_AUTH, properties)
     end
 
+    def check_concurrent_sessions
+      return unless existing_user.is_a? User
+      return unless existing_user.decorate.too_many_sessions?
+
+      flash[:error] = t('errors.messages.concurrent_sessions')
+      redirect_to root_url
+    end
+
     def cache_active_profile
       cacher = Pii::Cacher.new(current_user, user_session)
       cacher.save(params[:user][:password])
     end
+
+    def track_user_session
+      session_syncer.clean(current_user)
+      IdentityLinker.new(current_user, Identity::LOCAL, session.id).link_identity
+    end
+
+    def session_syncer
+      SessionSyncer.new(session_store)
+    end
+
+    def session_store
+      session.instance_variable_get('@by')
```

Comment:
agreed. ugly. `session` is not from Devise though, it is Rails native. See http://www.rubydoc.info/docs/rails/4.1.7/ActionDispatch/Request/Session

The session store seems to be intentionally private, yet we need to be able to keep it in sync with our sessions table. It seemed preferable to do that using the existing session storage class rather than duplicating a bunch of code to connect to Redis, create key strings and check for their existence.